pkg/controller/pkg/manager/reconciler.go

```diff
@@ -0,0 +1,401 @@
+	np := func() v1alpha1.Package { return &v1alpha1.Configuration{} }
+	nr := func() v1alpha1.PackageRevision { return &v1alpha1.ConfigurationRevision{} }
+	nrl := func() v1alpha1.PackageRevisionList { return &v1alpha1.ConfigurationRevisionList{} }
+
+	r := NewReconciler(mgr,
+		WithInstallNamespace(namespace),
+		WithNewPackageFn(np),
+		WithNewPackageRevisionFn(nr),
+		WithNewPackageRevisionListFn(nrl),
+		WithLogger(l.WithValues("controller", name)),
+		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor("configuration"))),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Configuration{}).
+		Owns(&v1alpha1.ConfigurationRevision{}).
+		Complete(r)
+}
+
+// NewReconciler creates a new package reconciler.
+func NewReconciler(mgr ctrl.Manager, opts ...ReconcilerOption) *Reconciler {
+	r := &Reconciler{
+		client: resource.ClientApplicator{
+			Client:     mgr.GetClient(),
+			Applicator: resource.NewAPIPatchingApplicator(mgr.GetClient()),
+		},
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+
+	return r
+}
+
+// Reconcile package.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	p := r.newPackage()
+	if err := r.client.Get(ctx, req.NamespacedName, p); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		log.Debug(errGetPackage, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPackage)
+	}
+
+	// Get existing package revisions.
+	prs := r.newPackageRevisionList()
+	if err := r.client.List(ctx, prs, client.MatchingLabels(map[string]string{parentLabel: p.GetName()})); err != nil {
+		log.Debug(errListRevisions, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errListRevisions)
+	}
+
+	// Identify or create an install pod.
+	pod := &corev1.Pod{}
+
+	// NOTE(hasheddan): we have to create the pod to get the image ID of the
+	// package. If we move to parsing the image manifest from registry directly
+	// then we can instead create the package revision and let it manage the
+	// full lifecycle of its pod.
+	podName := strings.NewReplacer("/", "-", ":", "-").Replace(p.GetSource())
+	if err := r.client.Get(ctx, types.NamespacedName{Name: podName, Namespace: r.namespace}, pod); err != nil {
+		if kerrors.IsNotFound(err) {
+			pod = buildInstallPod(podName, r.namespace, p.GetSource(), map[string]string{parentLabel: p.GetName()}, p.GetPackagePullSecrets(), p.GetPackagePullPolicy())
+			meta.AddOwnerReference(pod, meta.AsOwner(meta.TypedReferenceTo(p, p.GetObjectKind().GroupVersionKind())))
+			if err := r.client.Create(ctx, pod); err != nil {
```

Comment:
@kasey brought up a good question this morning about gc of completed Pods, which I wanted to add some context around...

`Succeeded` pods will be garbage collected according to the [`terminated-pod-gc-threshold` flag](https://kubernetes.cn/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection) on the `kube-controller-manager`, which has a [default value of 12500](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/config/v1alpha1/defaults.go#L34). This means that (with this current implementation), if an install Pod gets garbage collected (or manually deleted for that matter), the package revision will report that it is unhealthy, which will trigger a reconcile from the package controller, which will see that the Pod does not exist, and will recreate it. The package revision controller will then be successful on its next read of the Pod logs, will update to healthy, and that will be updated in the package as well.

This is assuming the current package revision though. If an _old_ package revision pod gets garbage collected, that revision will become unhealthy. This isn't a _huge_ problem, but it is a problem because an old current revision does still check that it has owner references on all of its installed objects, so it will fail to do that and the package controller will not recreate its Pod.

One way to get around this is to read the list of `Objects` we have on the revision (that get populated after reading Pod logs and modifying the objects), and if the package revision is inactive we can just use those to set the owner references (we are just adding them to the current object in the API server, so we don't need to know what it looks like in the package).

That being said, besides the old package revision being unhealthy, the fact that it is unhealthy is likely to not have effect on the system because nothing should be removing the owner references it has on an object, so it won't be updating them, but the object should still be garbage collected when the revision is deleted.

But don't worry, there's more complexity! If someone was to do a backup & restore of their cluster using something like Velero, the UID of the old package revision could change, and if it was the only owner reference on an object, that object would be GC'ed because it would see that none of its owners exist, and the Pod would not be created for the old package revision so it would not be able to go and update the owner references.

> aside: if there is no finalizer on the objects created by the package revision controller, could they be garbage collected on a backup and restore in the time it takes for the owner / controller references to get updated, even for the current package revision?

One thing that this all reaffirms in my mind is that storing objects in pod logs after initial unpack is a less than ideal situation. I hope that we can gain traction in a future iteration to unpack packages and either store them in volumes or some other medium.

Reply:
In regards to that last comment, another thing I would be in support of is getting the hash in the package controller (using something like `go-containerregistry`), but then deferring all unpacking and management of pods to the package revision controller.